lib/async_hooks.js

```diff
@@ -207,6 +208,95 @@ class AsyncResource {
   }
 }
 
+function getResource(id) {
+  const weak = idResourcePairing.get(id);
+  if (!weak) {
+    return undefined;
+  }
+  return weak.get();
+}
+
+function getCurrentResource() {
+  const id = executionAsyncId();
+  return getResource(id);
+}
+
+// AsyncContext part
+const idResourcePairing = new Map();
+const contexts = new Set();
+const asyncContextHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    // save new resource
+    idResourcePairing.set(asyncId, new WeakReference(resource));
+    contexts.forEach((context) => {
+      context._propagate(resource, getResource(triggerAsyncId));
+    });
+  },
+  destroy(asyncId) {
+    idResourcePairing.delete(asyncId);
+  },
+  promiseResolve(asyncId) {
+    idResourcePairing.delete(asyncId);
+  }
+});
+
+class AsyncContext {
+
+  constructor() {
+    // WeakMap resource -> store
+    this._resourceStore = new WeakMap();
+  }
+
+  /**
+   * propagate the context from a parent resource to a child one
```

Comment:
```suggestion
   * Propagate the context from a parent resource to a child one.
```

I'll stop mentioning it, but can you use proper capitalization and punctuation in the comments.